File: 프로그래머스/1/42889. 실패율/실패율.py

def solution(N, stages):
    answer = {}

    total_players = len(stages)

    for i in range(1, N+1):
        challengers = stages.count(i)
        if total_players == 0:
            answer[i] = 0
            continue
        
        answer[i] = challengers / total_players

        total_players -= challengers

    return sorted(answer, key=answer.get, reverse=True)

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;
import javafx.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int s : stages) {
            map.put(s, map.getOrDefault(s, 0) + 1);
        }
        
        int total = stages.length;
        List<Pair<Double, Integer>> list = new ArrayList<>();
        for (int i = 1; i <= N; i++) {
            list.add(new Pair<>(total == 0 ? 0 : (double)map.getOrDefault(i, 0) / total, i));
            total -= map.getOrDefault(i, 0);
        }
        
        list.sort((a, b) -> a.getKey().equals(b.getKey()) 
                  ? a.getValue() - b.getValue() 
                  : b.getKey().compareTo(a.getKey()));
        
        for (int i = 0; i < N; i++) {
            answer[i] = list.get(i).getValue();
        }
        
        return answer;
    }
}